File: prettier-plugin-csharpier/src/Printer/Types/ParenthesizedLambdaExpression.ts
import { Doc } from "prettier";
import { PrintMethod } from "../PrintMethod";
import { SyntaxToken, SyntaxTreeNode } from "../SyntaxTreeNode";
import { concat, group, hardline, indent, join, softline, line, doubleHardline } from "../Builders";
import { BlockNode, printBlock } from "./Block";
import { ParameterListNode, printParameterList } from "./ParameterList";

export interface ParenthesizedLambdaExpressionNode extends SyntaxTreeNode<"ParenthesizedLambdaExpression"> {
    // modifiers: SyntaxToken[]; appears to duplicate async
    parameterList?: ParameterListNode;
    arrowToken?: SyntaxToken;
    block?: BlockNode;
    expressionBody?: SyntaxTreeNode;
    asyncKeyword?: SyntaxToken;
    // body?: SyntaxTreeNode; appears to duplicate block
}

export const printParenthesizedLambdaExpression: PrintMethod<ParenthesizedLambdaExpressionNode> = (path, options, print) => {
    const node = path.getValue();
    const parts: Doc[] = [];
    if (node.asyncKeyword) {
        parts.push("async ");
    }
    parts.push(path.call(o => printParameterList(o, options, print), "parameterList"), " => ");
    if (node.expressionBody) {
        parts.push(path.call(print, "expressionBody"));
    } else if (node.block) {
        parts.push(path.call(o => printBlock(o, options, print), "block"));
    }

    return concat(parts);
};

File: Src/CSharpier/SyntaxPrinter/SyntaxNodePrinters/ParenthesizedLambdaExpression.cs
using System.Diagnostics;

namespace CSharpier.SyntaxPrinter.SyntaxNodePrinters;

internal static class ParenthesizedLambdaExpression
{
    public static Doc Print(ParenthesizedLambdaExpressionSyntax node, FormattingContext context)
    {
        return Doc.Concat(PrintHead(node, context), PrintBody(node, context));
    }

    public static Doc PrintHead(ParenthesizedLambdaExpressionSyntax node, FormattingContext context)
    {
        return Doc.Concat(
            AttributeLists.Print(node, node.AttributeLists, context),
            Modifiers.PrintSorted(node.Modifiers, context),
            node.ReturnType != null
                ? Doc.Concat(Node.Print(node.ReturnType, context), " ")
                : Doc.Null,
            ParameterList.Print(node.ParameterList, context),
            " ",
            Token.Print(node.ArrowToken, context)
        );
    }

    public static Doc PrintBody(ParenthesizedLambdaExpressionSyntax node, FormattingContext context)
    {
        return node.Body switch
        {
            BlockSyntax block => Doc.Concat(
                block.Statements.Count > 0 ? Doc.HardLine : " ",
                Block.Print(block, context)
            ),
            _ => Doc.Group(Doc.Indent(Doc.Line, Node.Print(node.Body, context))),
        };
    }
}
